pkg/v2/tkr/webhook/cluster/vsphere-template-resolver/template/resolver.go

```diff
@@ -0,0 +1,355 @@
+	templateQuery, err := createTemplateQueryFromTKRData(tkrDataValue, topologyVersion)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "error while building control plane query")
+	}
+
+	if templateQuery != nil {
+		cpQuery[*templateQuery] = struct{}{}
+		cpTKRDatas[templateQuery] = cpTKRData
+	}
+
+	return cpQuery, cpTKRDatas, nil
+}
+
+// mdDataValue holds the TKR Data and the constructed Template Query for a single Machine Deployment.
+type mdDataValue struct {
+	TKRData       *resolver_cluster.TKRData
+	TemplateQuery *templateresolver.TemplateQuery
+}
+
+// getMDQueryAndData: returns a template query and a map of the index of the TKR_DATA as the key, and mdDataValue as the value.
+// This works under the following assumptions:
+// - Every MD contains exactly one TKR_DATA.
+// - Every TKR_DATA contains exactly one TKRDataValue matching the topology's k8sVersion.
+// - Thus, every TKR_DATA can only map to one query.
+func getMDQueryAndData(cluster *clusterv1.Cluster) (map[templateresolver.TemplateQuery]struct{}, map[int]*mdDataValue, error) {
+	// Get TKR Data variable for each MD and populate query.
+	topology := cluster.Spec.Topology
+	if topology.Workers == nil {
+		return nil, nil, nil
+	}
+	topologyVersion := cluster.Spec.Topology.Version
+
+	mdData := map[int]*mdDataValue{}
+
+	mdQuery := map[templateresolver.TemplateQuery]struct{}{}
+	workers := topology.Workers
+	if workers != nil {
+		mds := workers.MachineDeployments
+		for i := 0; i < len(mds); i++ {
+			var mdTKRData resolver_cluster.TKRData
+			err := util_topology.GetMDVariable(cluster, i, varTKRData, &mdTKRData)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "error parsing TKR_DATA machine deployment %v", mds[i].Name)
+			}
+
+			// Collect the TKR_DATA so it can be used later for processing the result.
+			tkrDataValue, ok := (mdTKRData)[topologyVersion]
+			if !ok {
+				// No tkr data value for topology version, append an empty query
+				continue
+			}
+
+			// Build the query with OVA/OS details
+			templateQuery, err := createTemplateQueryFromTKRData(tkrDataValue, topologyVersion)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "error while building machine deployment query for machine deployment %v", mds[i].Name)
+			}
+
+			if templateQuery != nil {
+				mdQuery[*templateQuery] = struct{}{}
+				mdData[i] = &mdDataValue{TKRData: &mdTKRData, TemplateQuery: templateQuery}
+			}
+		}
+	}
+	return mdQuery, mdData, nil
+}
+
+func createTemplateQueryFromTKRData(tkrDataValue *resolver_cluster.TKRDataValue, topologyVersion string) (*templateresolver.TemplateQuery, error) {
+	ovaVersion, ok := tkrDataValue.OSImageRef[keyOSImageVersion].(string)
+	if !ok {
+		return nil, fmt.Errorf("ova version is invalid or not found for topology version %v", topologyVersion)
+	}
+
+	templatePath, ok := tkrDataValue.OSImageRef[keyOSImageTemplate].(string)
+	if ok && len(templatePath) > 0 {
+		// No resolution needed as template Path already exists.
```

Comment:
If the TKR_DATA has already a resolution to a template for given K8s version, it would ignore the resolution request. Since the users can move the template to a different path, we should have a comment here on what the user is supposed to do and also need to document this.

Reply:
We do need to document this.

Reply:
Added this in the release notes:

> * If the template is moved (the template path is changed) and the cluster is already resolved, users will need to manually edit the cluster and remove the corresponding `osImageRef.template` in the TKR Data.

Reply:
Thanks,  Please add a comment in the code and also a small note on why we choose this. This would help later if it needs to be updated.

Reply:
Done!

```go
// Users will need to explicitly remove the template related labels and re-trigger.
// This is necessary to prevent unnecessary template resolution every time there is an update to the cluster.
// Template resolution is seen as an expensive operation
```